### ISSUE TITLE
NewlineAtEndOfFile=lf does not create error when file uses crlf

### ISSUE THREAD

**Issue by jayvdb** [checkstyle/checkstyle#4074]
http://checkstyle.sourceforge.net/config_misc.html#NewlineAtEndOfFile

NewlineAtEndOfFile only checks the end of the file for bytes matching the EOL type , for the length of the EOL type.

As `lf` is only one byte, it does not detect that the previous byte `cr` before the `lf`.
Therefore a `crlf` combination at EOF is accepted as a valid match for `lineSeparator=lf`

```
$ cat TestConfig.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="NewlineAtEndOfFile">
    <property name="lineSeparator" value="lf"/>
  </module>
</module>

$ echo -ne "1\r\n" > newline.txt 

$ hd newline.txt 
00000000  31 0d 0a                                          |1..|
00000003


$ java -jar checkstyle-7.7-all.jar -c TestConfig.xml newline.txt 
Starting audit...
Audit done.
```

Expected: violation as last line is "0d0a" but "0a"(lf) is expected.

**Comment by Vladlis** [comment]
I guess `lf` should not cover neither `crlf` nor `lfcr`.

**Comment by jayvdb** [comment]
Sorry my original issue description wrote `lfcr` was allowed; I have corrected that to be `crlf`.

**Comment by rnveach** [comment]
@romani This issue looks valid to me.
We accidentally hit something like this in https://github.com/checkstyle/checkstyle/pull/4075#issuecomment-288479685

File was changed to `\r\n` for last line, and both travis and appveyor passed.
IMO, travis should have failed because it would expect `\n` alone.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/NewlineAtEndOfFileCheck.java#L59 is Fileset Check so could be launched on any file type.
So all before required symbol is not our business.
But some people might not care, probably majority.
So we need make new option that will recheck no end of line  control characters before symbol.

@rnveach , @jayvdb do you agree on this ?
I am ok to have this property activated by default to follow majority.

**Comment by rnveach** [comment]
> will recheck no end of line control characters before symbol.

So if I configure the check to have files end with `\n` and the file has `\n\n\n` at the end, this will be a violation? or will it check for every variation except the one in use?

**Comment by romani** [comment]
My previous request for new option was bad idea, if new line is requested to be last it mean that previous was also a line that have its end.
Cases where user use checkstyle not on text files we are not considering a correct usage of checkstyle - we will not support this.

So what we have:
```
"crlf" (Windows-style), \r\n
"cr" (Mac-style), \r
"lf" (Unix-style) and \n
"lf_cr_crlf" (lf, cr or crlf). any - so \n is ok even \r\n is requested is fine.
```

Lets see possible use-cases:
```
config: lineSeparator='\r'

file: 
\r --> OK
\r\n --> Violation
\n --> Violation

-----------------------------

config: lineSeparator='\n'

file: 
\r --> Violation
\r\n --> Violation ?! (issue case, need to check for "\r")
\n --> ok

-----------------------------

config: lineSeparator='\r\n'

file: 
\r --> Violation
\r\n --> ok
\n --> Violation
```

Problem is that in Check we do read last 1 or 2 chracters only, base on length of lineSeparator.
So we need to improve Check and read lineSeparator.lenght * 2 . Attention that lineSeparator could be 'system' and user can put whatever in it.
So we need to expect two line end symbols to be present in in this chunk (from previous line and on last)

expected behavior base on new logic:
```
config: lineSeparator='\r'

file: 
\r\r -- we read \r\r --> OK
\r\n\r\n -- we read \r\n --> Violation
\n\n -- we read \n\n --> Violation

-----------------------------

config: lineSeparator='\n'

file: 
\r\r -- we read \r\r --> Violation
\r\n\r\n -- we read \r\n --> Violation !!!! (issue case)
\n\n -- we read \n\n --> ok

-----------------------------

config: lineSeparator='\r\n'

file:
\r\r -- we read ....\r\r --> Violation
\r\n\r\n -- we read \r\n\r\n --> ok
\n\n -- we read ....\n\n --> Violation
```

@jayvdb , @rnveach , Am I make sense now ?

**Comment by rnveach** [comment]
Fix was merged
